Workaround for bad definitions bug in PCL 1.8.1

schornakj opened the pull request:
Fix for #9. This change reverts to use `add_definitions` instead, which is unfortunately very un-modern-CMake.

schornakj committed:
use add_definitions instead of target_compile_definitions with PCL_DEFINITIONS

gavanderhoorn commented:
What about removing the empty `-D` from the `PCL_DEFINITIONS` (conditionally on whether PCL 1.8.x is used of course)?

Could be as simple as a `list(REMOVE_ITEM PCL_DEFINITIONS "-D")` (but I haven't checked whether that touches any of the other defs).

schornakj commented:
When I use `message(WARNING ${PCL_DEFINITIONS})` I get this output:

```
CMake Warning at /home/mrtd/other_workspaces/yak_melodic_ws/src/yak/yak/CMakeLists.txt:65 (message):
   -DDISABLE_DAVIDSDK-DDISABLE_DSSDK-DDISABLE_PCAP-DDISABLE_PNG-DDISABLE_LIBUSB_1_0
```

I think this shows that the contents of `${PCL_DEFINITIONS}` are well-formed, and that the bad definition is being introduced specifically when I use `target_compile_definitions(${PROJECT_NAME}_frontend PUBLIC ${PCL_DEFINITIONS})`.

Levi-Armstrong commented:
In 1.8.1 they combine both compiler definition and compiler options into PCL_DEFINITIONS which prevents the use of target_compile_definitions and target_compile_options. In tesseract I have a [script](https://github.com/ros-industrial-consortium/tesseract/blob/e2f43729de30b6b495f9c0969f262fd70757111e/tesseract/tesseract_urdf/CMakeLists.txt#L20-L43) that splits this into definitions and options so I could use it with targets. This is fixed int 1.9.1 where you now have PCL_DEFINITIONS and PCL_COMPILER_OPTIONS, but they also export targets so you really don't need them.

schornakj committed:
Add a macro to sanitize and parse lists of options and definitions

schornakj committed:
Switch method for adding PCL_DEFINITIONS on PCL version

schornakj commented:
I added a modified version of @Levi-Armstrong 's script as a macro. I also added a check to use `target_compile_definitions` and `target_compile_options` the "normal" way for PCL 1.9 and newer. This builds for me on both PCL 1.8.1 and 1.9.1.
